cudax/include/cuda/experimental/__launch/hierarchy_draft.cuh

```diff
@@ -0,0 +1,428 @@
+ * configuration will be replaced with concrete dimensions if they are meta dimensions or passed through otherwise. This
+ * function needs to take the stream and the function, because the finalization of some meta dimensions require
+ * informations about the device or the kernel. Compared to the function pointer overloads, this version also needs to
+ * know the types of arguments of the kernel functor, which are taken in the first template argument pack.
+ *
+ * @tparam Args
+ * Types of arguments of the kernel functor
+ *
+ * @param stream
+ * Stream that is going to be used in the launch function to launch kernel with this configuration or dimensions
+ *
+ * @param conf_or_dims
+ * Configuration or dimensions that are going to be finalized
+ *
+ * @param kernel
+ * Kernel functor that the configuration are intended for
+ */
+
+_CCCL_DIAG_PUSH
+_CCCL_DIAG_SUPPRESS_MSVC(4180) // qualifier applied to function type has no meaning; ignored in is_function
+template <typename... Args,
+          typename Kernel,
+          typename ConfOrDims,
+          typename = ::cuda::std::enable_if_t<!::cuda::std::is_function_v<std::remove_pointer_t<Kernel>>>>
+_CCCL_NODISCARD auto finalize(::cuda::stream_ref stream, const ConfOrDims& conf_or_dims, const Kernel&)
+{
+  return finalize(
+    stream, conf_or_dims, detail::get_kernel_launcher<Kernel, finalized_t<ConfOrDims>, as_kernel_arg_t<Args>...>());
+}
+_CCCL_DIAG_POP
+
+/**
+ * @brief Returns a hierarchy or a configuration with the hierarchy updated to replace meta dimsnions with concrete
+ * dimensions
+ *
+ * This function will create a new kernel_configuration or hierarchy_dimensions based on what is passed in. It will be
+ * finalized for the passed in kernel functor. Each level in the hierarchy or the hierarchy contained in the
+ * configuration will be replaced with concrete dimensions if they are meta dimensions or passed through otherwise. This
+ * function needs to take the stream and the function, because the finalization of some meta dimensions require
+ * informations about the device or the kernel. Compared to the function pointer overloads, this version also needs to
+ * know the types of arguments of the kernel functor, which are taken as the last arguments pack
+ *
+ * @param stream
+ * Stream that is going to be used in the launch function to launch kernel with this configuration or dimensions
+ *
+ * @param conf_or_dims
+ * Configuration or dimensions that are going to be finalized
+ *
+ * @param kernel
+ * Kernel functor that the configuration are intended for
+ *
+ * @param args
+ * Arguments that the kernel functor will be launched with
+ */
+template <typename... Args, typename Kernel, typename ConfOrDims>
+_CCCL_NODISCARD auto finalize(
+  ::cuda::stream_ref stream, const ConfOrDims& conf_or_dims, const Kernel& kernel, [[maybe_unused]] const Args&... args)
+{
+  return finalize<Args...>(stream, conf_or_dims, kernel);
+}
+
```

Comment:
i think these `finalize` overloads will be more discoverable as a member function of the conf/dims.

Reply:
I changed them to be member functions. Originally `finalize` was ment to mimic `launch` structure with its order of arguments. Now that is a member function its no longer the case, so I changed the stream argument to a device argument. I had some issues with `finalize` taking a stream as first argument, because then the API looks like something happening asynchronously in that stream, where here its not the case. Now with it taking a device I don't think its a concern and if someone needs to convert it, there is `stream.device()`

Reply:
Can we move the definition back into the class?

Reply:
I think we will run into circular dependencies with `configuration.cuh`